### PR TITLE
[PD-2212]  Remove jQuery from User Management

### PR DESCRIPTION
**Pull request by darrint**
* Create a common API class useful from any app.
* Use it from Manage Users and Reporting.
* Remove use of `/* global $ */` in Manage Users.

We depend on a common behavior in jQuery implemented by it's param function starting in jQuery 1.4.

http://api.jquery.com/jquery.param/

This affects sending arrays and objects as form data. jQuery renames array keys to `field[]` and object keys to `field[key]`.

Using this I can use the same code for all browsers.

### CODE REVIEW

File: gulp/src/common/myjobs-api.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+  async ajaxWithFormData(method, url, data) {
+    const formData = param(data);
+    const response = await fetch(url, {
+      method: method,
+      body: formData,
+      credentials: 'include',
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+        'X-CSRFToken': this.csrf,
```
Comment (darrint) [inline gulp/src/common/myjobs-api.js]: This header works for all http methods in my testing. This eliminates the need to modify request bodies to include the csrf token.

File: gulp/src/manageusers/Role.jsx [hunk 2]
```diff
@@ -83,10 +44,49 @@ class Role extends React.Component {
       activites: currentActivitiesObject,
     });
   }
-  handleSaveRoleClick() {
+  async initialApiLoad() {
```
Comment (darrint) [inline gulp/src/manageusers/Role.jsx]: This reads more clearly with whitespace differences hidden.

File: gulp/src/manageusers/Role.jsx [hunk 3]
```diff
@@ -30,43 +27,7 @@ class Role extends React.Component {
...
-        this.setState({
-          apiResponseHelp: '',
-          roleName: '',
-          availableUsers: roleObject.available_users,
-          assignedUsers: [],
-          activities: activities,
-        });
-      }.bind(this));
-    }
+    this.initialApiLoad();
```
Comment (darrint) [inline gulp/src/manageusers/Role.jsx]: The linter is overzealous about use of setState in componentDidMount. It correctly allows it in callbacks but does not know that any code after an `await` is equivalent to code in a callback. So I move the api call to a separate method. The linter is also wants the new method moved below event handlers.

So this change is much smaller than it appears. It also reads more clearly with whitespace differences hidden. `?w=1`

File: gulp/src/common/myjobs-api.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  checkStatus(response) {
+    if (response.status === 200) {
+      return response;
+    }
+
+    const error = new Error(response.statusText);
+    error.response = response;
+    throw error;
```
Comment (darrint) [inline gulp/src/common/myjobs-api.js]: Note that I'm treating failed requests as exceptions. See below for how the exceptions handlers look.

File: gulp/src/manageusers/Role.jsx [hunk 5]
```diff
@@ -218,27 +216,18 @@ class Role extends React.Component {
...
-          this.setState({
-            apiResponseHelp: 'Role not deleted. Insufficient privileges.',
-          });
-        }
-      }.bind(this));
+    try {
+      await api.delete( '/manage-users/api/roles/delete/' + roleId + '/');
+      await this.props.callRolesAPI();
+      history.pushState(null, '/roles');
+    } catch (e) {
```
Comment (darrint) [inline gulp/src/manageusers/Role.jsx]: Here's how api error handling looks.

File: gulp/src/manageusers/Role.jsx [hunk 6]
```diff
@@ -1,11 +1,8 @@
-/* global $ */
-
 import React from 'react';
 import _ from 'lodash-compat';
 import {Link} from 'react-router';
 import Button from 'react-bootstrap/lib/Button';
 
-import {getCsrf} from 'common/cookie';
```
Comment (darrint) [inline gulp/src/manageusers/Role.jsx]: Users of api are no longer concerned with csrf. This is encapsulated in the api class.

File: gulp/src/reporting/api.js [hunk 7]
```diff
@@ -4,153 +4,87 @@
...
-          'Content-Type': 'application/x-www-form-urlencoded',
-        },
-      });
-      return this.parseJSON(this.checkStatus(response));
-    }
+  async postToReportingApi(url, data) {
+    return this.myJobsApi.post(url, data);
+  }
 
-    async listReports() {
```
Comment (darrint) [inline gulp/src/reporting/api.js]: The linter is not enforcing indentation level between classes and functions. I went ahead and fixed this file. This difference is best viewed with whitespace differences hidden.

File: gulp/package.json [hunk 8]
```diff
@@ -37,6 +37,7 @@
     "tinycolor2": "~1.3.0",
     "warning": "^2.1.0",
     "webpack": "^1.12.9",
-    "lodash-compat": "~3.10.1"
+    "lodash-compat": "~3.10.1",
+    "jquery-param": "~0.2.0"
```
Comment (darrint) [inline gulp/package.json]: jquery-param provides jQuery like form data serialization that we want. It has no dependencies.

**Comment by aspidites** [comment]
I thought there was a big push to get *rid* of jQuery? This seems to be adding jQuery-specific dependencies. Note, I don't mind jQuery, I would just like us to be consistent about our attitude toward one set of librires over another. If we are keeping jQuery, then I would think that fetch is redundant.

**Comment by darrint** [comment]
What the library provides is equivalent to jQuery's param function. It is not jQuery. We're likely indirectly using `$.param` a lot in current ajax calls. Manage users definitely was. Having the myjobs-api.js calls work this way makes for fewer differences to worry about when calling other existing APIs.

https://www.npmjs.com/package/jquery-param

**Comment by aspidites** [comment]
I see. Never mind then.

**Comment by tdphillips** [comment]
What does this give us that we didn't have before?

File: gulp/src/manageusers/Role.jsx [hunk 9]
```diff
@@ -198,19 +198,17 @@ class Role extends React.Component {
...
         });
       }
-    }.bind(this));
+    }
   }
-  handleDeleteRoleClick() {
-    const history = this.props.history;
+  async handleDeleteRoleClick() {
+    const {api, history} = this.props;
     // Temporary until I replace $.ajax jQuery with vanilla JS ES6 arrow function
```
Comment (JLMcLaugh) [inline gulp/src/manageusers/Role.jsx]: Is this comment obsolete now?

**Comment by darrint** [comment]
@tdphillips 

* Details of making API requests such as csrf token handling are encapsulated in `class MyJobsApi` instead of copy-pasted everywhere we make requests.
* All browsers serialize request bodies the same way.
* The existing jQuery style parameter serialization is preserved.
* Some IE8 ugliness that was in `reporting/api.js` is gone. i.e. `ancientSerialize` is gone.
* No need to pass `/* global $ */` in from the window object.

File: gulp/src/manageusers/Role.jsx (hunk 9 above)
Comment (darrint) [inline gulp/src/manageusers/Role.jsx]: Thanks.

File: gulp/src/manageusers/Role.jsx (hunk 5 above)
Comment (aspidites) [inline gulp/src/manageusers/Role.jsx]: I would expect to see:
```javascript
await this.props.callDeleteAPI(roleID);
await this.props.callRolesAPI();
```
or this:
```javascript
await api.delete(...);
await api.get(...);
```
But not a mixture of both. Any particular reason for this asymmetry?

Comment (darrint) [inline gulp/src/manageusers/Role.jsx]: If you trace down this.props.callRolesAPI (and it's sibling, callUsersAPI) it's a fairly involved function that I would like to refactor. It is arguably in the wrong place as well. I started fixing that even but decided not to in order to keep this PR from getting too big. In this PR those functions stop using jQuery and I leave it at that.

I created this ticket to continue to refactor them: [PD-2228](https://directemployersfoundation.atlassian.net/browse/PD-2228)

**Pull request merged**